### ISSUE TITLE
Web App constantly using one core even without any requests

### ISSUE THREAD

**Issue by Jyothsnasrinivas** [typelead/eta#839]
As monnef (https://gitlab.com/monnef) pointed out, Eta is constantly using one core even if there are no requests for a simple json api using scotty. Looks like a runtime bug.

https://gitlab.com/monnef/czech-work-days-api/blob/e895687d735b02ea0b85b16a02dd16d337e6b657/src/Main.hs

**Comment by shirouto** [comment]
@rahulmutt I can confirm that [https://github.com/typelead/eta/commit/cf2b252431cebc56ad2159e78b168366e4b839f9](cf2b252) does fix the stack trace above.

However, while testing that I ran into another error:
```java.lang.IllegalArgumentException: Cannot dereference 0x47001000: Exceeded higher bound of the address space.
        at eta.runtime.storage.ManagedHeap.throwIllegalAddressException(ManagedHeap.java:103)
        at eta.runtime.storage.ManagedHeap.getBlock(ManagedHeap.java:88)
        at eta.runtime.io.MemoryManager.getBlockSlow(MemoryManager.java:102)
        at eta.runtime.io.MemoryManager.getBlock(MemoryManager.java:96)
        at eta.runtime.io.MemoryManager.getBoundedBuffer(MemoryManager.java:115)
        at eta.base.Utils.c_read(Utils.java:148)
        at base.ghc.io.FD$sat_sGZZ9F.applyV(FD.hs:436)
        at base.foreign.c.Error.throwErrnoIfMinus1RetryMayBlock2(Error.hs:342)
        at base.ghc.io.FD.$wa4(FD.hs:433)
        at warp.network.wai.handler.warp.Recv.$wa2(Recv.hs)
        at warp.network.wai.handler.warp.Recv.$wa3(Recv.hs)
        at warp.network.wai.handler.warp.Run$sat_s5M2T.applyV(Run.hs)
        at warp.network.wai.handler.warp.Run.serveConnection1(Run.hs)
        at warp.network.wai.handler.warp.Run$sat_s5M13.apply1V(Run.hs)
        at eta.runtime.apply.PAP1_1.applyV(PAP1_1.java:19)
        at eta.runtime.thunk.Thunk.applyV(Thunk.java:120)
        at eta.runtime.exception.Exception.unmaskAsyncExceptions(Exception.java:91)
        at base.control.exception.Base$sat_sGIRN.applyV(Base.hs)
        at eta.runtime.exception.Exception.catch_(Exception.java:129)
        at base.control.exception.Base$sat_sGIS0.applyV(Base.hs:262)
        at eta.runtime.exception.Exception.maskAsyncExceptions(Exception.java:42)
        at base.control.exception.Base.bracket1(Base.hs:264)
        at warp.network.wai.handler.warp.Run$sat_s5M14.applyV(Run.hs)
        at eta.runtime.exception.Exception.unmaskAsyncExceptions(Exception.java:91)
        at warp.network.wai.handler.warp.Settings$sat_s2GFC.apply1V(Settings.hs)
        at eta.runtime.apply.PAP1_1.applyV(PAP1_1.java:19)
        at eta.runtime.thunk.Thunk.applyV(Thunk.java:120)
        at eta.runtime.exception.Exception.maskAsyncExceptions(Exception.java:42)
        at base.control.exception.Base$sat_sGIR5.applyV(Base.hs)
        at eta.runtime.exception.Exception.catch_(Exception.java:129)
        at base.control.exception.Base.bracket1(Base.hs:262)
        at warp.network.wai.handler.warp.Run$sat_s5M16.apply1V(Run.hs)
        at eta.runtime.apply.PAP1_1.applyV(PAP1_1.java:19)
        at eta.runtime.thunk.Thunk.applyV(Thunk.java:120)
        at eta.runtime.exception.Exception.maskAsyncExceptions(Exception.java:42)
        at base.control.exception.Base$sat_sGIR5.applyV(Base.hs)
        at eta.runtime.exception.Exception.catch_(Exception.java:129)
        at base.control.exception.Base.bracket1(Base.hs:262)
        at warp.network.wai.handler.warp.Run$sat_s5M17.applyV(Run.hs)
        at eta.runtime.exception.Exception.catch_(Exception.java:129)
        at warp.network.wai.handler.warp.Run$sat_s5M18.apply1V(Run.hs)
        at eta.runtime.apply.PAP1_1.applyV(PAP1_1.java:19)
        at eta.runtime.thunk.Thunk.applyV(Thunk.java:120)
        at eta.runtime.exception.Exception.catch_(Exception.java:129)
        at warp.network.wai.handler.warp.Settings$sat_s2GFE.applyV(Settings.hs)
        at eta.runtime.stg.Closures$EvalLazyIO.enter(Closures.java:108)
        at eta.runtime.stg.Capability.schedule(Capability.java:183)
        at eta.runtime.concurrent.WorkerThread.run(WorkerThread.java:16)
recv: failed```

At first sight, it seems to suggest that the jvm needs more heap space, but I am pretty sure there should be no memory leak in the testing code. Besides, while Warp seems to serve up local requests quite well (until the server chokes up), it often fails to do so with remote requests (it claims that "Client closed the connection prematurely", while the client gets an "empty response" --- I checked, and there is no significant latency though). It goes without saying, there are no such issues when I am building the same code with vanilla ghc (lts-6.27). I will look closer at the Warp issue in the following days.

**Comment by Jyothsnasrinivas** [comment]
@shirouto Thanks for looking into this! This doesn't seem to be warp-specific: https://github.com/typelead/eta/issues/845. Can you file distinct issues for each of the issues you discover?

**Comment by shirouto** [comment]
@Jyothsnasrinivas Yes, you are obviously right. I have updated the last comment to not include anything on that memory management issue. And I will open separate issues from now on when unsure.

**Comment by shirouto** [comment]
I went ahead and made an *extremely* naive implementation for the exponential backoff as indicated by @rahulmutt . The good news is that it does improve dramatically the Warp server response so that it no longer returns empty responses. Unfortunately, it doesn't seem to help much with the CPU load. Even in the absence of any requests and the polling being throttled up to 5 seconds, the Warp server still hugs the entire core and it does not let it go. However, as pointed out above, my implementation is naive and it is likely I may have missed some important detail. For example, I would expect that ```Concurrent.yield``` and ```Concurrent.threadWaitIO``` would need a more nuanced approach than a standard implementation otherwise appropriate for ```MVar``` handling, for example.

I will follow up on the thread count story in a separate ticket.

**Comment by rahulmutt** [comment]
@shirouto Great to see you investigate these - just a heads up `eta-build rts-clean` is super fast way of re-installing the modified runtime. Feel free to submit a work-in-progress PR and I can give inline comments.

**Comment by shirouto** [comment]
Thanks for the tip. Have been running ```stack exec eta-build -- install rts-clean``` instead, which was a mouthful. I guess ```eta-build``` does not need any stack path.

I think I identified the issue for the poor CPU management in the Warp eta-rts combination: the ```Finalizer``` thread starts taking over after one request and keeps remaining that way even without receiving any additional requests, being responsible for the CPU drain. The number of threads created is not that large (below 18 threads per request, with a large churn, but that seems to be the way Warp typically runs from what I can tell). So I guess somewhere between the Warp and the eta-rts, a lot of resources are not manually deallocated, putting extra burden on the finalizer.

Between this and the issue @rahulmutt has pointed out about over-polling, it makes for a pretty stifling experience for the CPU. 

Sorry for messing things up while investigating this. I have been taking a lot of wrong paths.

**Comment by rahulmutt** [comment]
Oh that assumes you did `stack install eta-build` beforehand. `stack exec eta-build -- install rts-clean` also works.

We do allocate a number of weak references to link `Ptr`'s (used to implement `ByteString` and `Text`) to their corresponding off-heap memory address managed by `eta.runtime.io.MemoryManager`. Maybe you can try looking at the number of weak references and how that fluctuates when a request is sent? I do recall reading that a sufficiently large number of weak references can cause issues.

The churn of threads is probably because the Warp as you're using it roughly allocates a thread per `forkIO`. @ouromoros's [GSoC work](https://github.com/ouromoros/eta-gsoc) was to use Fibers to implement it and he observed at least a 2x improvement in number of requests handled per second.

**Comment by rahulmutt** [comment]
Hmm but if the `ByteString`/`Text`'s were properly deallocated after a request, all the weak references should be discharged! This means some of the `ByteString`/`Text`'s are not being discharged as they should (maybe a memory leak somewhere)?

**Comment by shirouto** [comment]
Well, I counted the ```WeakReference```s for memory resources, and I did not notice anything fishy. The garbage collector is timely disposing of them, without any sudden increase in their number.

However, it seems the Finalizer thread chokes on releasing some thread...

```
"Finalizer" #3 daemon prio=8 os_prio=0 tid=0x00007f12b008c000 nid=0x327 runnable [0x00007f129251e000]
   java.lang.Thread.State: RUNNABLE
        at java.lang.Thread.yield(Native Method)
        at eta.runtime.util.MPSCLongQueue.allocateNode(MPSCLongQueue.java:107)
        at eta.runtime.util.MPSCLongQueue.write(MPSCLongQueue.java:94)
        at eta.runtime.storage.Block.sendFreeMessage(Block.java:77)
        at eta.runtime.storage.ManagedHeap.attemptFree(ManagedHeap.java:95)
        at eta.runtime.io.MemoryManager.free(MemoryManager.java:80)
        at sun.reflect.GeneratedMethodAccessor1.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at eta.runtime.stg.JavaFinalizer.finalize(JavaFinalizer.java:26)
        at java.lang.System$2.invokeFinalize(System.java:1270)
        at java.lang.ref.Finalizer.runFinalizer(Finalizer.java:102)
        at java.lang.ref.Finalizer.access$100(Finalizer.java:34)
        at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:217)
```
I sampled multiple times on multiple runs, and each time the ```Finalizer``` takes over it hanging in that native call to ```java.lang.Thread.yield()```. So I started looking for any misbehaving thread, and I ran into something like this:

```
"Thread-1" #10 prio=5 os_prio=0 tid=0x00007fb889672000 nid=0x6fc5 runnable [0x00007fb8346d4000]
   java.lang.Thread.State: RUNNABLE
        at sun.nio.ch.IOUtil.drain(Native Method)
        at sun.nio.ch.EPollSelectorImpl.doSelect(EPollSelectorImpl.java:104)
        - locked <0x00000006c573d840> (a java.lang.Object)
        at sun.nio.ch.SelectorImpl.lockAndDoSelect(SelectorImpl.java:86)
        - locked <0x00000006c573d618> (a sun.nio.ch.Util$3)
        - locked <0x00000006c573d608> (a java.util.Collections$UnmodifiableSet)
        - locked <0x00000006c573d4f0> (a sun.nio.ch.EPollSelectorImpl)
        at sun.nio.ch.SelectorImpl.selectNow(SelectorImpl.java:105)
        at eta.runtime.concurrent.Concurrent.checkForReadyIO(Concurrent.java:387)
        at eta.runtime.stg.Capability.idleLoop(Capability.java:555)
        at eta.runtime.stg.Capability.blockedLoop(Capability.java:602)
        at eta.runtime.stg.Capability.schedule(Capability.java:166)
        at eta.runtime.concurrent.WorkerThread.run(WorkerThread.java:16)
```
Well, the hanging call may be one of the ```sun.nio.ch.IOUtil.drain()```, ```sun.nio.ch.EPollArrayWrapper.epollWait()```, or ```sun.nio.ch.EPollArrayWrapper.interrupt()```, but all originate in the ```eta.runtime.concurrent.Concurrent.checkForReadyIO()``` call. 

So this seems a tad... disappointing: is this a JVM implementation issue? I am on a OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode). 

Otherwise, while testing this I have been using a stripped down version of the app: it barely creates three ```Text``` values, so I cannot imagine that would be a factor. I even swapped that for a version without any ```isString``` instance, with similar results: the ```checkForReadyIO``` starts overtaking the CPU, then eventually the ```Finalizer``` issue picks up too, and in combination they account for two cores. 

Yes, I have been using ```forkIO```, and for my use case I never had any problem. This time might be worth looking into fibers though. Thanks for the nudge.

**Comment by rahulmutt** [comment]
The `selectNow()` calls is probably because the server is idle and waiting for another connection. But exponential backoff should have reduced the effect of this right? In which method did you implement it?

**Comment by rahulmutt** [comment]
@shirouto Can you host the code you're testing somewhere so I can attempt to reproduce it?

**Comment by shirouto** [comment]
@rahulmutt  Even with the exponential backoff I am running into this issue. But again, I may have have missed the point entirely on that implementation: I did submit a PR, so you can test/critique.

Also, in order to test it, I used different code versions. I include herein the one that has the lowest LOC count.

```
{-# language OverloadedStrings #-}

module Main where

import Web.Scotty
import Data.Monoid (mconcat)
import Network.Wai.Handler.Warp
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import System.IO (stderr)


serverOpts :: Options
serverOpts = Options 1 $
  setOnException (\_ e -> TIO.hPutStrLn stderr $ T.pack $ show e) $
  setHost "0.0.0.0" $
  setPort 6001 $
  defaultSettings

main :: IO ()
main = scottyOpts serverOpts $
  get "/:word" $ do
    beam <- param "word"
    html $ mconcat ["<h1>Scotty, ", beam, " me up!</h1>"]
```

Also, the cabal file which is also straight out of the etlas init:

```
-- Initial warp_test.cabal generated by etlas init.  For further 
-- documentation, see http://etlas.typelead.com/users-guide

name:                warp_test
version:             0.1.0.0
license:             GPL-2
license-file:        LICENSE
author:              shirouto
build-type:          Simple
extra-source-files:  ChangeLog.md
cabal-version:       >=1.10

executable warp_test
  main-is:             Main.hs
  build-depends:       base >=4.8 && <4.9
                     , scotty
                     , warp
                     , text
  hs-source-dirs:      src
  default-language:    Haskell2010
```

In order to test it, the surest method I found is to simply ```yes "curl http://localhost:6001" | sh``` and wait 10-15 seconds. Once it hangs up, the CPU usage stabilizes around 100%-200% and it never goes down no matter if there are no further requests.

Thank you a lot for your help.

**Comment by rahulmutt** [comment]
@shirouto Thank you for the great reproduction case! Was able to reproduce the finalizer problem with pretty much the same stack trace and it looks to be one of the causes of eating up too much CPU. The other cause is the `checkForReadyIO` as you mentioned and as I mentioned in the PR, I'll see if I can convert it to a non-polling mechanism (more like an event loop).

**Comment by rahulmutt** [comment]
I fixed the finalizer issue (a bug in the concurrent data structure we use to signal freeing memory) and that brought down the average processor usage from around 200% to 100%. Will continue fixing the remaining issues by reducing the amount of polling we do and update this thread accordingly.

**Comment by rahulmutt** [comment]
This is now fixed on master after a refactoring of the thread scheduling in the runtime. CPU usage is now reasonable. Fix will be included in the upcoming v0.8.6b2 release.
